xmake/core/package/package.lua

```diff
@@ -1711,13 +1726,34 @@ end
 function _instance:find_tool(name, opt)
     opt = opt or {}
     self._find_tool = self._find_tool or sandbox_module.import("lib.detect.find_tool", {anonymous = true})
-    return self._find_tool(name, {cachekey = opt.cachekey or "fetch_package_system",
-                                  installdir = self:installdir({readonly = true}),
-                                  version = true, -- we alway check version
-                                  require_version = opt.require_version,
-                                  norun = opt.norun,
-                                  system = opt.system,
-                                  force = opt.force})
+    local fetchopt = {
+        cachekey = opt.cachekey or "fetch_package_system",
+        installdir = installdir,
+        version = true, -- we alway check version
+        require_version = opt.require_version,
+        norun = opt.norun,
+        system = opt.system,
+        force = opt.force
+    }
+
+    local installdir = self:installdir({readonly = true})
+    fetchopt.installdir = installdir
+    local fetchinfo = self._find_tool(name, fetchopt)
+    if fetchinfo then
+        return fetchinfo
+    end
+
+    -- try from global install dir (if package has install_locally policy)
+    installdir = self:installdir({readonly = true, ignorelocal = true, ignorecache = true})
```

Comment:
why? we should not find twice.

Reply:
we need only modify is_local

Reply:
we need to find twice because the package may be installed globally.

for example if I have libsdl already installed, and I'm using package.install_locally on a project using libsdl and other packages, I only want the other packages to be installed in the project folder. Otherwise it's not different from just setting XMAKE_PKG_INSTALLDIR to another value.

only changing package:is_local() means xmake-repo fetch won't work as it uses the installdir to fetch the package manifest.

Reply:
If libsdl has been installed to a global path, then its `is_local()` should return false instead of looking twice.

Reply:
yeah but how do you know if the package has been installed to XMAKE_PKG_INSTALLDIR without first trying to find it?

This is a project policy, not a per-package policy.

Reply:
we can improve `package:installdir()` to support `package:installdir({local = true})`

```lua
function _instance:installdir(...)
    -- TODO opt
    local installdir = self._INSTALLDIR
    if not installdir then
        installdir = self:get("installdir")
        if not installdir then
            local name = self:name():lower():gsub("::", "_")
            if opt.local or self:is_local() then.  <----
```

then check if `package:manifest_file()`  and `package:manifest_file({local = true})` exists in self:is_local()

Reply:
shouldn't `package:is_local()` should return true if the policy is in effect? because yeah we can force it but we would need to do the same in _instance:cachedir() and _instance:buildir() (which are testing package:is_local() for now)

Reply:
so if we set local policy globally, is_local is always true. and xmake should always install to local, even if it has been installed globally. we should also reinstall it to local. do not twice find.

Reply:
this will make xmake reinstall all packages already installed in XMAKE_PKG_INSTALLDIR to the project. this is no different from setting XMAKE_PKG_INSTALLDIR to the project folder.

this policy aims to reduce disk space usage by allowing to test xmake projects on a separate drive than the main one (where XMAKE_PKG_INSTALLDIR points to) and be able to remove a project and all the dependencies it installed quickly.  
Making xmake reinstall already installed packages will just make it lose build time and space.

Reply:
> shouldn't should return true if the policy is in effect? 

If we just want to incrementally install packages locally, then as I just said, we just need to determine if the global manifest file exists inside is_local, and if it does, and we're still using the globally installed packages, then the current instance of the package is still a global package, and is_local will return false even though we've set up a local policy.

_instance:cachedir() and _instance:buildir()  will be changed automaticall when is_local() return true or false

Reply:
There are also some problems with calling manifest_file inside is_local. is_local will be called too early, when the buildhash is still uncertain, and this will generate some incorrect installdir.

I don't think this feature is well implemented. The current patch causes is_local to return true, but it may still be using the library files in the global directory. This inconsistent state is also unacceptable, and it causes twice find_tool/package.

I think priority should be given to maintainability, consistency of package state. If the implementation of this feature is hard to maintain and be implemented, I have to disregard it for now.

Reply:
Another idea would be to let is_local untouched and add is_local_install (which would check policy), which would be checked by :installdir()/:cachedir()/:buildir().

The main issue is to make xmake try to find the package in the project folder and then in the XMAKE_PKG_INSTALLDIR, since xmake-repo package finder tries to use :installdir()

Reply:
I know, but I also need to consider the complexity and maintainability of the code implementation.

There is no implementation that satisfies me at the moment and I need some time. When I have time I will think about how to implement it better. But right now I'm busy with other issues and xpack.